File: 907-koko-eating-bananas/koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    let left = 0
    let right = Math.max(...piles)

    while (left < right) {
        const mid = Math.floor((left + right) / 2)

        if (canEatAll(mid, piles, h)) {
            right = mid
        } else {
            left = mid + 1
        }
    }

    return left
};

function canEatAll(k, piles, h) {
    let hours = 0

    for (const pile of piles) {
        hours += Math.ceil(pile/ k)
        if (hours > h) return false
    }
    return true
}

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low, high = 1, max(piles)

        while low < high:
            mid = (low + high) // 2
            if self.canFinish(piles, mid, h):
                high = mid
            else:
                low = mid + 1

        return low

    def canFinish(self, piles, mid, h):
        hours = 0

        for pile in piles:
            hours += (pile + mid - 1) // mid

            if hours > h:
                return False

        return hours <= h